### ISSUE TITLE
Source Code for Self Learning data array for Cartpole

### ISSUE THREAD

**Issue by Deadsg** [meta-introspector/https-streamlit.io-community-llm-hackathon-2023#41]
import gym
import numpy as np

class QLearningAgent:
    def __init__(self, num_actions, learning_rate=0.1, discount_factor=0.9, exploration_prob=0.2):
        self.num_actions = num_actions
        self.learning_rate = learning_rate
        self.discount_factor = discount_factor
        self.exploration_prob = exploration_prob
        self.q_table = np.zeros((num_states, num_actions))

    def choose_action(self, state):
        if np.random.rand() < self.exploration_prob:
            return np.random.choice(self.num_actions)
        else:
            return np.argmax(self.q_table[state])

    def update_q_table(self, state, action, reward, next_state):
        best_next_action = np.argmax(self.q_table[next_state])
        target = reward + self.discount_factor * self.q_table[next_state, best_next_action]
        self.q_table[state, action] += self.learning_rate * (target - self.q_table[state, action])

class SelfLearningDataArray:
    def __init__(self):
        self.observations = []  # List to store state observations
        self.actions = []       # List to store actions taken
        self.rewards = []       # List to store rewards received

    def add_data(self, observation, action, reward):
        self.observations.append(observation)
        self.actions.append(action)
        self.rewards.append(reward)

    def get_batch(self):
        return np.array(self.observations), np.array(self.actions), np.array(self.rewards)

    def clear_data(self):
        self.observations = []
        self.actions = []
        self.rewards = []

# Initialize the CartPole environment
env = gym.make('CartPole-v1')

# Initialize Q-learning agent and SelfLearningDataArray
num_actions = env.action_space.n
agent = QLearningAgent(num_actions)
data_array = SelfLearningDataArray()

# Training loop
for episode in range(num_episodes):
    state = env.reset()
    done = False

    while not done:
        action = agent.choose_action(state)
        next_state, reward, done, _ = env.step(action)

        # Add data to the data array
        data_array.add_data(state, action, reward)

        state = next_state

    # Train the agent with the collected data
    observations_batch, actions_batch, rewards_batch = data_array.get_batch()
    for i in range(len(observations_batch)):
        agent.update_q_table(observations_batch[i], actions_batch[i], rewards_batch[i], observations_batch[i+1])

    # Clear the data array for the next set of episodes
    data_array.clear_data()
